torch/_logging/_internal.py

```diff
@@ -738,9 +741,40 @@ def _default_formatter():
 DEFAULT_FORMATTER = _default_formatter()
 
 
+class TorchLogsFilter(logging.Filter):
+    def __init__(self, ranks):
+        self.ranks = set(ranks)
+        super().__init__()
+
+    def filter(self, record) -> bool:
+        if dist.is_available() and dist.is_initialized():
+            return dist.get_rank() in self.ranks
+
+        return True
+
+
+def _default_filter() -> Optional[TorchLogsFilter]:
+    ranks_str = os.environ.get(LOG_RANK_ENV_VAR, None)
+    if ranks_str is None:
+        return None
+
+    ranks = [int(rank_str) for rank_str in ranks_str.split(",")]
+    if not ranks:
+        raise ValueError(
+            "Expected comma separated list of int (e.g. 0,1,2,3), received {ranks_str}"
+        )
+
+    return TorchLogsFilter(ranks)
+
+
+DEFAULT_FILTER = _default_filter()
+
+
 def _setup_handlers(create_handler_fn, log):
     debug_handler = _track_handler(create_handler_fn())
     debug_handler.setFormatter(DEFAULT_FORMATTER)
+    if DEFAULT_FILTER:
```

Comment:
don't we always go into this if branch as `DEFAULT_FILTER` is a global variable? maybe we can remove the if clause if that's the case.

Reply:
I think there's some additional overhead to run each log through a no-op filter, so I only attach the filter when it is needed